framework/src/main/java/org/checkerframework/framework/type/QualifierHierarchy.java

```diff
@@ -164,20 +238,71 @@ default boolean isSubtype(
    */
   // The fact that null is returned if the qualifiers are not in the same hierarchy is used by the
   // collection version of LUB below.
-  @Nullable AnnotationMirror leastUpperBound(
+  protected abstract @Nullable AnnotationMirror leastUpperBoundQualifiers(
       AnnotationMirror qualifier1, AnnotationMirror qualifier2);
 
+  /**
+   * Returns the least upper bound (LUB) of the qualifiers {@code qualifier1} and {@code
+   * qualifier2}. Returns {@code null} if the qualifiers are not from the same qualifier hierarchy.
+   * Ignores Java basetypes.
+   *
+   * <p>Examples:
+   *
+   * <ul>
+   *   <li>For NonNull, leastUpperBound('Nullable', 'NonNull') &rArr; Nullable
+   * </ul>
+   *
+   * @param qualifier1 the first qualifier; may not be in the same hierarchy as {@code qualifier2}
+   * @param qualifier2 the second qualifier; may not be in the same hierarchy as {@code qualifier1}
+   * @return the least upper bound of the qualifiers, or {@code null} if the qualifiers are from
+   *     different hierarchies
+   */
+  // The fact that null is returned if the qualifiers are not in the same hierarchy is used by the
+  // collection version of LUB below.
+  public final @Nullable AnnotationMirror leastUpperBoundQualifiersOnly(
+      AnnotationMirror qualifier1, AnnotationMirror qualifier2) {
+    return leastUpperBoundQualifiers(qualifier1, qualifier2);
+  }
+
+  /**
+   * Returns the least upper bound (LUB) of the qualifiers {@code qualifier1} and {@code
+   * qualifier2}. Returns {@code null} if the qualifiers are not from the same qualifier hierarchy.
+   *
+   * <p>Examples:
+   *
+   * <ul>
+   *   <li>For NonNull, leastUpperBound('Nullable', 'NonNull') &rArr; Nullable
+   * </ul>
+   *
+   * @param qualifier1 the first qualifier; may not be in the same hierarchy as {@code qualifier2}
+   * @param tm1 the type on which qualifier1 appears
+   * @param qualifier2 the second qualifier; may not be in the same hierarchy as {@code qualifier1}
+   * @param tm2 the type on which qualifier2 appears
+   * @return the least upper bound of the qualifiers, or {@code null} if the qualifiers are from
+   *     different hierarchies
+   */
+  // The fact that null is returned if the qualifiers are not in the same hierarchy is used by the
+  // collection version of LUB below.
+  public @Nullable AnnotationMirror leastUpperBoundShallow(
+      AnnotationMirror qualifier1, TypeMirror tm1, AnnotationMirror qualifier2, TypeMirror tm2) {
+    return leastUpperBoundQualifiers(qualifier1, qualifier2);
```

Comment:
This doesn't use the passed type mirrors.  Does the following make sense?
```suggestion
    boolean tm1IsRelevant = atypeFactory.isRelevant(tm1);
    boolean tm2IsRelevant = atypeFactory.isRelevant(tm2);
    if (tm1IsRelevant == tm2IsRelevant) {
      return leastUpperBoundQualifiers(qualifier1, qualifier2);
    } else if (tm1IsRelevant) {
      return qualifier1;
    } else { // if(tm2IsRelevant) {
      return qualifier2;
    }
```